### PR TITLE
feature(whl): add continuous bc

### PR DESCRIPTION
**Pull request by kxzxvbk**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: run_bc.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    state_dict = expert_policy.collect_mode.state_dict()
+    collect_config = [deepcopy(bipe_sac_config), deepcopy(bipe_sac_create_config)]
+    # collect_demo_data(
+    #     collect_config, seed=0, state_dict=state_dict, expert_data_path=expert_data_path, collect_count=collect_count
+    # )
+    collect_episodic_demo_data(
+        deepcopy(collect_config), seed=0, state_dict=state_dict, expert_data_path=expert_data_path, collect_count=collect_count
+    )
+    collect_episodic_demo_data(
+        collect_config, seed=0, state_dict=state_dict, expert_data_path='eval_data.pkl', collect_count=10,
```
Comment (Will-Nie) [inline run_bc.py:85]: repetitive function

File: run_bc.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    )
```
Comment (Will-Nie) [inline run_bc.py:86]: Why do you collect twice?

File: ding/policy/bc.py [hunk 3]
```diff
@@ -195,3 +195,194 @@ def _get_train_sample(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
+
+    def _monitor_vars_learn(self):
+        return ['cur_lr', 'total_loss', 'forward_time', 'backward_time', 'sync_time']
+
+    def _init_eval(self):
+        self._eval_model = model_wrap(self._model, wrapper_name='base')
+        self._eval_model.reset()
+
+    def _forward_eval(self, data):
+        if self.cfg.eval.evaluator.cfg_type == 'MetricSerialEvaluatorDict' or isinstance(data, torch.Tensor):
```
Comment (Will-Nie) [inline ding/policy/bc.py:286]: What if you only have one evaluator_env in your config? Is there going to be a bug?

File: ding/entry/serial_entry_bc.py [hunk 4]
```diff
@@ -60,9 +61,13 @@ def serial_pipeline_bc(
...
     learner.call_hook('before_run')
     stop = False
-
     for epoch in range(cfg.policy.learn.train_epoch):
         # Evaluate policy performance
+        loss_list = []
+        for _, bat in enumerate(eval_loader):
+            res = policy._forward_eval(bat['obs'])
+            loss_list.append(torch.nn.L1Loss()(res['action'], bat['action']).item())
+        tb_logger.add_scalar("validation loss", sum(loss_list) / len(loss_list), epoch)
```
Comment (Will-Nie) [inline ding/entry/serial_entry_bc.py:70]: Unifying your x-axis

File: ding/policy/bc.py [hunk 5]
```diff
@@ -195,3 +195,194 @@ def _get_train_sample(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
+        # self._mse_loss = nn.MSELoss()
+
+    def _forward_learn(self, data):
+        if not isinstance(data, dict):
+            data = default_collate(data)
+        if self._cuda:
+            data = to_device(data, self._device)
+        self._learn_model.train()
+        with self._timer:
+            if self.cfg.eval.evaluator.cfg_type == 'MetricSerialEvaluatorDict':
```
Comment (Will-Nie) [inline ding/policy/bc.py:250]: What's the reason to put eval config into inside learn

File: run_bc.py (hunk 2 above)
Comment (kxzxvbk) [reply run_bc.py:86]: the second time of collecting is used for validation

File: ding/model/template/__init__.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 from .qac import QAC, DiscreteQAC
 from .pdqn import PDQN
 from .vac import VAC
-from .bc import BC
+from .bc import BC, ContinuousBC
```
Comment (PaParaZz1) [inline ding/model/template/__init__.py:6]: rename to `DiscreteBC` and `ContinuousBC`

File: ding/model/template/bc.py [hunk 7]
```diff
@@ -94,3 +96,270 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+                after each FC layer, if ``None`` then default set to ``nn.ReLU()``.
+            - norm_type (:obj:`Optional[str]`): The type of normalization to after network layer (FC, Conv), \
+                see ``ding.torch_utils.network`` for more details.
+        """
+        super(ContinuousBC, self).__init__()
+        obs_shape: int = squeeze(obs_shape)
+        action_shape = squeeze(action_shape)
+        self.action_shape = action_shape
+        self.action_space = action_space
+        assert self.action_space in ['regression', 'reparameterization', 'hybrid']
```
Comment (PaParaZz1) [inline ding/model/template/bc.py:150]: remove hybrid and update following comments

File: ding/model/template/bc.py [hunk 8]
```diff
@@ -94,3 +96,270 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+                    actor_head_hidden_size,
+                    action_shape.action_type_shape,
+                    actor_head_layer_num,
+                    activation=activation,
+                    norm_type=norm_type,
+                )
+            )
+            self.actor = nn.ModuleList([actor_action_type, actor_action_args])
+
+        self.twin_critic = twin_critic
```
Comment (PaParaZz1) [inline ding/model/template/bc.py:203]: remove twin_critic

File: ding/model/template/bc.py [hunk 9]
```diff
@@ -94,3 +96,270 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+                    critic_head_hidden_size,
+                    1,
+                    critic_head_layer_num,
+                    final_tanh=False,
+                    activation=activation,
+                    norm_type=norm_type
+                )
+            )
+
+    def forward(self, inputs: Union[torch.Tensor, Dict[str, torch.Tensor]], mode: str) -> Dict[str, torch.Tensor]:
```
Comment (PaParaZz1) [inline ding/model/template/bc.py:237]: remove `compute_actor` and `compute_critic`, just a `forward`

File: ding/policy/bc.py [hunk 10]
```diff
@@ -195,3 +195,194 @@ def _get_train_sample(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
+        #    if epoch <= self._cfg.learn.warmup_epoch:
+        #        return self._cfg.learn.warmup_lr / self._cfg.learn.learning_rate
+        #    else:
+        #        ratio = epoch // self._cfg.learn.decay_epoch
+        #        return math.pow(self._cfg.learn.decay_rate, ratio)
+        #self._lr_scheduler = LambdaLR(self._optimizer, lr_scheduler_fn)
+        #self._lr_scheduler.step()
+        self._learn_model = model_wrap(self._model, 'base')
+        self._learn_model.reset()
+        # self._mse_loss = nn.SmoothL1Loss()
```
Comment (PaParaZz1) [inline ding/policy/bc.py:239]: add a loss switch here and use config to control

File: ding/policy/bc.py [hunk 11]
```diff
@@ -195,3 +195,194 @@ def _get_train_sample(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
+
+    def _init_collect(self) -> None:
+        r"""
+        Overview:
+            Collect mode init method. Called by ``self.__init__``.
+            Init traj and unroll length, collect model.
+            Enable the eps_greedy_sample
+        """
+        self._unroll_len = self._cfg.collect.unroll_len
+        #self._gamma = self._cfg.discount_factor  # necessary for parallel
```
Comment (PaParaZz1) [inline ding/policy/bc.py:323]: remove these commented code

File: run_bc.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
+from ding.entry import serial_pipeline_bc, serial_pipeline, collect_demo_data
```
Comment (PaParaZz1) [inline run_bc.py:1]: simplify this file and move it to dizoo as a example, such as `mujoco_bc_main.py`

File: ding/model/template/bc.py [hunk 13]
```diff
@@ -94,3 +96,270 @@ def forward(self, x: torch.Tensor) -> Dict:
         x = self.encoder(x)
         x = self.head(x)
         return x
+
+
+@MODEL_REGISTRY.register('continuous_bc')
+class ContinuousBC(nn.Module):
```
Comment (PaParaZz1) [inline ding/model/template/bc.py:102]: add unittest

File: ding/entry/serial_entry_bc.py [hunk 14]
```diff
@@ -18,6 +18,7 @@ def serial_pipeline_bc(
         seed: int,
         data_path: str,
         model: Optional[torch.nn.Module] = None,
+        max_iter=1e6,
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_bc.py:21]: int(1e6)

File: ding/entry/serial_entry_bc.py [hunk 15]
```diff
@@ -33,6 +34,8 @@ def serial_pipeline_bc(
         - policy (:obj:`Policy`): Converged policy.
         - convergence (:obj:`bool`): whether il training is converged
     """
+    cont = input_cfg[0].policy.type == 'continuous_bc'
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_bc.py:37]: use `continuous`, full name is OK

File: ding/entry/serial_entry_bc.py [hunk 16]
```diff
@@ -50,7 +53,12 @@ def serial_pipeline_bc(
     # Main components
     tb_logger = SummaryWriter(os.path.join('./{}/log/'.format(cfg.exp_name), 'serial'))
     dataset = NaiveRLDataset(data_path)
-    dataloader = DataLoader(dataset, cfg.policy.learn.batch_size, collate_fn=lambda x: x)
+    print('Dataset length' + str(len(dataset)))
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_bc.py:56]: use logging

File: ding/entry/serial_entry_bc.py [hunk 17]
```diff
@@ -60,15 +68,27 @@ def serial_pipeline_bc(
     # ==========
     learner.call_hook('before_run')
     stop = False
-
+    iter_cnt = 0
     for epoch in range(cfg.policy.learn.train_epoch):
         # Evaluate policy performance
+        loss_list = []
+        for _, bat in enumerate(eval_loader):
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_bc.py:75]: also add discrete viz, like acc

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/bc.py [hunk 18]
```diff
@@ -1,10 +1,12 @@
...
+from easydict import EasyDict
 
 from ding.torch_utils import get_lstm
 from ding.utils import MODEL_REGISTRY, SequenceType, squeeze
-from ..common import FCEncoder, ConvEncoder, DiscreteHead, DuelingHead, MultiHead
+from ..common import FCEncoder, ConvEncoder, DiscreteHead, DuelingHead, \
+        MultiHead, RegressionHead, ReparameterizationHead
 
 
 @MODEL_REGISTRY.register('bc')
```
Comment (PaParaZz1) [inline ding/model/template/bc.py:12]: rename to discrete bc

File: ding/model/template/bc.py [hunk 19]
```diff
@@ -94,3 +96,105 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+            - norm_type (:obj:`Optional[str]`): The type of normalization to after network layer (FC, Conv), \
+                see ``ding.torch_utils.network`` for more details.
+        """
+        super(ContinuousBC, self).__init__()
+        obs_shape: int = squeeze(obs_shape)
+        action_shape = squeeze(action_shape)
+        self.action_shape = action_shape
+        self.action_space = action_space
+        assert self.action_space in ['regression', 'reparameterization']
+        if self.action_space == 'regression':  # DDPG, TD3
```
Comment (PaParaZz1) [inline ding/model/template/bc.py:148]: remove comment DDPG, TD3

File: ding/policy/bc.py [hunk 20]
```diff
@@ -44,18 +48,15 @@ def _init_learn(self):
...
-        #        return self._cfg.learn.warmup_lr / self._cfg.learn.learning_rate
-        #    else:
-        #        ratio = epoch // self._cfg.learn.decay_epoch
-        #        return math.pow(self._cfg.learn.decay_rate, ratio)
-        #self._lr_scheduler = LambdaLR(self._optimizer, lr_scheduler_fn)
-        #self._lr_scheduler.step()
         self._learn_model = model_wrap(self._model, 'base')
         self._learn_model.reset()
-        self._ce_loss = nn.CrossEntropyLoss()
+        if not self._cfg.continuous:
```
Comment (PaParaZz1) [inline ding/policy/bc.py:53]: reverse if-else branch like this, such as:
```python
if self._cfg.continuous:
    ...
else:
    ...
```

File: ding/policy/bc.py [hunk 21]
```diff
@@ -104,7 +112,10 @@ def _forward_eval(self, data):
                 data = to_device(data, self._device)
             self._eval_model.eval()
             with torch.no_grad():
-                output = self._eval_model.forward(data)
+                if not self._cfg.continuous:
```
Comment (PaParaZz1) [inline ding/policy/bc.py:115]: why still some codes about metric evaluator here, you should modify it like this:
```python
if self._cfg.metric.eval_type == 'metric':
  ...
elif self._cfg.metric.eval_type == 'interaction':
  ...
else:
  raise KeyError
```

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/bc.py (hunk 18 above)
Comment (kxzxvbk) [reply ding/model/template/bc.py:12]: I merge continous bc and discrete bc into one policy bc. How can I name it?

File: dizoo/mujoco/example/mujoco_bc_main.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    cfg = compile_config(cfg, seed=seed, env=env_fn, auto=True, create_cfg=create_cfg, save_cfg=True)
+    # Create main components: env, policy
+    if env_setting is None:
+        env_fn, collector_env_cfg, evaluator_env_cfg = get_vec_env_setting(cfg.env)
+    else:
+        env_fn, collector_env_cfg, evaluator_env_cfg = env_setting
+    collector_env = create_env_manager(cfg.env.manager, [partial(env_fn, cfg=c) for c in collector_env_cfg])
+    evaluator_env = create_env_manager(cfg.env.manager, [partial(env_fn, cfg=c) for c in evaluator_env_cfg])
+    collector_env.seed(cfg.seed)
+    evaluator_env.seed(cfg.seed, dynamic_seed=False)
```
Comment (PaParaZz1) [inline dizoo/mujoco/example/mujoco_bc_main.py:44]: we don't need codes related to env when loading policy

File: ding/policy/bc.py [hunk 23]
```diff
@@ -44,18 +48,15 @@ def _init_learn(self):
...
-        #self._lr_scheduler = LambdaLR(self._optimizer, lr_scheduler_fn)
-        #self._lr_scheduler.step()
         self._learn_model = model_wrap(self._model, 'base')
         self._learn_model.reset()
-        self._ce_loss = nn.CrossEntropyLoss()
+        if self._cfg.continuous:
+            if self._cfg.loss_type == 'l1_loss':
+                self._loss = nn.L1Loss()
+            elif self._cfg.loss_type == 'mse_loss':
+                self._loss = nn.MSELoss()
```
Comment (PaParaZz1) [inline ding/policy/bc.py:57]: add else branch add raise error

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]
